core/blockchain.go

```diff
@@ -476,19 +492,27 @@ func (bc *BlockChain) SetHead(head uint64) error {
 			bc.currentFastBlock.Store(newHeadFastBlock)
 			headFastBlockGauge.Update(int64(newHeadFastBlock.NumberU64()))
 		}
-	}
+		head := bc.CurrentBlock().NumberU64()
 
+		// If setHead underflown the freezer threshold and the block processing
+		// intent afterwards is full block importing, delete the chain segment
+		// between the stateful-block and the sethead target.
+		var wipe bool
+		if head+1 < frozen {
+			wipe = pivot == nil || head >= *pivot
```

Comment:
If we rewind below the pivot(e.g. from the synced status), how can we reenable the fast sync?

Reply:
In this case the `HeadBlock` is genesis again, but we should change the `handler.fastSync` to 1 again.

Reply:
Ah, I wanted to do that, but then saw a code segment that reenabled fast sync automatically. I was wrong however, that code only runs on the eth.ProtocolManager construction, not later during sync initiation.

Reply:
PTAL